src/modules/ekf2/EKF/zero_innovation_heading_update.cpp

```diff
@@ -0,0 +1,81 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2022 PX4 Development Team. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file zero_innovation_heading_update.cpp
+ * Control function for ekf heading update when at rest or no other heading source available
+ */
+
+#include "ekf.h"
+
+void Ekf::controlZeroInnovationHeadingUpdate()
+{
+	// When at rest or no source of yaw aiding is active yaw fusion is run selectively to enable yaw gyro
+	// bias learning when stationary on ground and to prevent uncontrolled yaw variance growth
+	if (_control_status.flags.vehicle_at_rest && _control_status.flags.tilt_align) {
+
+		const float euler_yaw = wrap_pi(getEulerYaw(_R_to_earth));
+
+		// record static yaw when transitioning to at rest
+		if (!PX4_ISFINITE(_last_static_yaw)) {
+			_last_static_yaw = euler_yaw;
+		}
+
+		// fuse last static yaw at a limited rate (every 200 milliseconds)
+		if (isTimedOut(_time_last_heading_fuse, (uint64_t)200'000)) {
```

Comment:
I'm not sure if we should really always fuse a constant heading when a yaw aiding source if active. We've seen that the GNSS yaw data is sometimes resetting after the initial lock. This fixed yaw fusion would prevent the filter to re-converge to the new correct measurements

Reply:
> In that GNSS yaw case was the reset explicitly handled?

Not explicitly, it goes naturally through the fusion process

Reply:
> The 200 ms time is also a bit arbitrary and could maybe be increased quite a bit?

I cannot tell, it would require some testing

Reply:
> For the timeout here I used _time_last_heading_fuse with the intent that it's going to be skipped entirely when a healthy aid source is active.

Ok

> The 200 ms time is also a bit arbitrary and could maybe be increased quite a bit?

We could probably reduce the rate and the observation variance to obtain a similar result, yes, but 5Hz isn't so fast, is it?

Reply:
> but 5Hz isn't so fast, is it?

I was thinking of it in terms of timeout relative to last successful yaw fusion, but I suppose even that is potentially faster than GPS yaw. Maybe it would be better to do this explicitly from control_status flags and then specifically think about the no yaw aiding case (rate and observation variance).

Reply:
> but I suppose even that is potentially faster than GPS yaw

Yes, that's a problem, we shouldn't start to fuse artificial data between two GPS yaw samples.

> Maybe it would be better to do this explicitly from control_status flags and then specifically think about the no yaw aiding case (rate and observation variance).

Yes, why not doing like fake pos fusion? https://github.com/PX4/PX4-Autopilot/blob/92adc714247ae598eab9c39c000314473c323c00/src/modules/ekf2/EKF/fake_pos_control.cpp#L43-L49
It can only start if there's no other source of similar aiding and will fuse fake measurements at a limited rate